P5 - Replace levelDB with badgerDB in shed

jmozah opened the pull request:
This PR replaces leveldb with badgerdb.
Refer P5 in #66

janos commented:
This PR should be rebased with master when https://github.com/ethersphere/bee/pull/74 is merged. PR #74 has only one comment that needs to be resolved.

janos commented:
PR https://github.com/ethersphere/bee/pull/74 is merged, please, rebase.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	if err != nil {
+		db.logger.Errorf("error while counting keys in DB")
+		db.logger.Debugf("error while doing count. Error : %w", err.Error())
+		db.metrics.TotalFailCount.Inc()
```
Should this block return and avoid the Trace log line below?

jmozah replied:
no. I want ed to print the count if trace is enabled.

janos replied:
Should the count be irrelevant if there is an error?

jmozah replied:
ok. Add this in an else

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	if err != nil {
+		db.logger.Errorf("error doing count from. prefix : %v", string(prefix))
+		db.logger.Debugf("error while doing count from. prefix : %v, Error : %w", string(prefix), err.Error())
+		db.metrics.CountPrefixFailCount.Inc()
```
Should this block return and avoid the Trace log line below?

jmozah replied:
no. I wanted to print the count if trace is enabled.

janos replied:
Should the count be irrelevant if there is an error?

jmozah replied:
add this in an else.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		i := txn.NewIterator(o)
+		defer i.Close()
+
+		for i.Seek(prefix); i.ValidForPrefix(prefix); i.Next() {
```
This will iterate over all keys with prefixes to get the last one. Looks expensive.

Shed in swarm uses more efficient approach to seek ahead and have only one Prev call afterwards https://github.com/ethersphere/swarm/blob/master/shed/index.go#L375. Why was that abandoned?

jmozah replied:
i also thought about this.. There is no prev() call in badger. i can raise a request for them. until then this is the only way.

janos replied:
Couldn't the `IteratorOptions.Reverse` option be used to get the Prev functionality?

jmozah replied:
did that exactly

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
 		return err
 	}
-	db.metrics.WriteBatchCounter.Inc()
+	txn.Discard()
```
Why calling discard? Txn.Commit() also calls Txn.Discard() internally to cleanup the transaction.

jmozah replied:
good catch. But calling it twice does not harm it anyway. I will remove it for brevity.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	close(db.quit)
-	return db.ldb.Close()
+// Close shut's down the badger DB.
+func (db *DB) Close(ctx context.Context) (err error) {
```
The context is not used. Why is it added?

jmozah replied:
if we want to cancel Close() if it takes too long or something in future...What do you say.. keep it or remove it...?

janos replied:
I do not see any reason to add an argument that is not used in the function body.

janos replied:
Please, remove the unused argument.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-func (db *DB) Close() (err error) {
-	close(db.quit)
-	return db.ldb.Close()
+// Close shut's down the badger DB.
```
Shouldn't be just `shuts`?

janos commented on pkg/shed/field_string.go:
```diff
@@ -65,6 +65,9 @@ func (f StringField) Put(val string) (err error) {
+func (f StringField) PutInBatch(batch *badger.Txn, val string) {
+	err := batch.Set(f.key, []byte(val))
+	if err != nil {
+		f.logger.Debugf("could not set values in PutInBatch")
```
The error is not handled, not even put in log message.

This is very dangerous. We should return this error in order to stop any other actions on the batch. This can be cause of data inconstancies.

jmozah replied:
yeah. I will add error..

janos commented on pkg/shed/field_struct.go:
```diff
@@ -67,12 +67,15 @@ func (f StructField) Put(val interface{}) (err error) {
-	batch.Put(f.key, b)
+	err = batch.Set(f.key, b)
+	if err != nil {
+		f.logger.Debugf("could not put values in batch. Error: %s", err.Error())
```
Why the error is not returned, just logged?

janos replied:
Do not log error and return it.

janos commented on pkg/shed/field_uint64.go:
```diff
@@ -67,16 +67,19 @@ func (f Uint64Field) Put(val uint64) (err error) {
+func (f Uint64Field) PutInBatch(batch *badger.Txn, val uint64) {
+	err := batch.Set(f.key, encodeUint64(val))
+	if err != nil {
+		f.logger.Debugf("could not set uint64 value in PutInBatch. Error : %s", err.Error())
```
The error is not handled!

This is very dangerous. We should return this error in order to stop any other actions on the batch. This can be cause of data inconstancies.

janos replied:
Do not log error and return it.

janos commented on pkg/shed/index.go:
```diff
@@ -162,32 +161,37 @@ func (f Index) Get(keyFields Item) (out Item, err error) {
+		badgerItem, err := txn.Get(key)
 		if err != nil {
+			if err == badger.ErrKeyNotFound {
+				f.logger.Debugf("error could not find key in fill. Error: %s", err.Error())
```
Do not log error and return error.

janos commented on pkg/shed/index.go:
```diff
@@ -162,32 +161,37 @@ func (f Index) Get(keyFields Item) (out Item, err error) {
+		var decodedItem Item
+		err = badgerItem.Value(func(val []byte) error {
+			decodedItem, err = f.decodeValueFunc(item, val)
+			if err != nil {
+				return err
+			}
+			return nil
```
Can be `return err`.

jmozah replied:
i think it is correct... it should be return nil

janos replied:
The if is not needed, the comment is meant for all 4 lines, including if.

janos commented on pkg/shed/index.go:
```diff
@@ -162,32 +161,37 @@ func (f Index) Get(keyFields Item) (out Item, err error) {
+			return nil
+		})
 		if err != nil {
 			f.logger.Debugf("error decofing keyfields in Fill . Error: %s", err.Error())
```
Do not log error and return error.

janos commented on pkg/shed/index.go:
```diff
@@ -208,22 +212,30 @@ func (f Index) Has(keyFields Item) (bool, error) {
 	for i, keyFields := range items {
 		key, err := f.encodeKeyFunc(keyFields)
 		if err != nil {
 			f.logger.Debugf("keyfields encoding error in HasMulti. Error: %s", err.Error())
```
Do not log error and return error.

janos commented on pkg/shed/index.go:
```diff
@@ -259,7 +271,11 @@ func (f Index) PutInBatch(batch *leveldb.Batch, i Item) (err error) {
-	batch.Put(key, value)
+	err = batch.Set(key, value)
+	if err != nil {
+		f.logger.Debugf("could not set values in batch. Error : %s", err.Error())
```
Why the error is not passed to the caller?

This may result in committing an incomplete batch.

janos replied:
Do not log error and return it.

janos commented on pkg/shed/index.go:
```diff
@@ -276,13 +292,16 @@ func (f Index) Delete(keyFields Item) (err error) {
-	batch.Delete(key)
+	err = batch.Delete(key)
+	if err != nil {
+		f.logger.Debugf("could not delete items in batch. Error: %s", err.Error())
```
Why the error is not passed to the caller?

This may result in committing an incomplete batch.

janos replied:
Do not log error and return it.

janos commented on pkg/shed/index.go:
```diff
@@ -322,102 +341,66 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
+// itemFromKeyValue returns the Item from the current iterator position.
 // If the complete encoded key does not start with totalPrefix,
-// leveldb.ErrNotFound is returned. Value for totalPrefix must start with
+// ErrNotFound is returned. Value for totalPrefix must start with
```
This comment change is not correct. It will return `badger.ErrKeyNotFound` not `ErrNotFound`.

janos commented on pkg/shed/index.go:
```diff
@@ -322,102 +341,66 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
-		it.Prev()
-	} else {
-		it.Last()
-	}
-
 	totalPrefix := append(f.prefix, prefix...)
-	return f.itemFromIterator(it, totalPrefix)
+	k, v, err := f.db.Last(totalPrefix)
```
This change replaces an efficient implementation with one very inefficient. Related to comment to `db.Last`.

jmozah replied:
As i said in the older commit.. there is no way to do a Previous() while doing a iteration. Let me see how else can i fix this.

janos replied:
Couldn't the `IteratorOptions.Reverse` option be used to get the Prev functionality?

jmozah replied:
fixed this.

janos commented on pkg/shed/metrics.go:
```diff
@@ -13,95 +13,193 @@ type metrics struct {
 			Subsystem: subsystem,
-			Name:      "put_count",
-			Help:      "Number of times the PUT operation is done.",
+			Name:      "db_db_open_count",
```
Every metric name now has a prefix `db_`, this is not needed as there already is a subsystem "shed" set. It would be nicer to remove `db_` prefix for every metric in this PR.

janos commented on pkg/shed/vector_uint64.go:
```diff
@@ -66,16 +66,19 @@ func (f Uint64Vector) Put(i, val uint64) (err error) {
+func (f Uint64Vector) PutInBatch(batch *badger.Txn, i, val uint64) {
+	err := batch.Set(f.indexKey(i), encodeUint64(val))
+	if err != nil {
+		f.logger.Debugf("could not put uint64 value in batch. Error : ", err.Error())
```
Again, error is not returned, which may result in data inconsistencies if the batch is committed, as the error is not propagated.

janos replied:
Do not log error and return it.

janos commented:
There is also a panic from the badger codebase which is related to Go 1.14 unsafe.Pointer safety rules. This relates to badger 1.6.0 used in this PR. I didi not check if this issue is present in version 2.0.3.

```
=== RUN   TestNewDB
fatal error: checkptr: unsafe pointer arithmetic

goroutine 16 [running]:
runtime.throw(0xe12690, 0x23)
	/opt/hostedtoolcache/go/1.14.2/x64/src/runtime/panic.go:1116 +0x72 fp=0xc0000893d0 sp=0xc0000893a0 pc=0x446cd2
runtime.checkptrArithmetic(0xc00022a080, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.14.2/x64/src/runtime/checkptr.go:43 +0xb5 fp=0xc000089400 sp=0xc0000893d0 pc=0x418115
github.com/AndreasBriese/bbloom.Bloom.JSONMarshal(0x0, 0xe, 0xc00022a080, 0x8, 0x8, 0x9, 0x1ff, 0x7, 0x37, 0x0, ...)
	/home/runner/go/pkg/mod/github.com/!andreas!briese/bbloom@v0.0.0-20190306092124-e2d15f34fcf9/bbloom.go:238 +0xd1 fp=0xc0000894e8 sp=0xc000089400 pc=0xa086d1
github.com/dgraph-io/badger/table.(*Builder).Finish(0xc0002028a0, 0xc0002800cb, 0x14, 0x533322e)
	/home/runner/go/pkg/mod/github.com/dgraph-io/badger@v1.6.0/table/builder.go:229 +0x7c1 fp=0xc000089ac8 sp=0xc0000894e8 pc=0xa0b621
github.com/dgraph-io/badger.writeLevel0Table(0xc0001aa060, 0x2a00000000, 0x39, 0x0, 0x0, 0x0, 0xf096a0, 0xc000206018, 0x0, 0x0)
	/home/runner/go/pkg/mod/github.com/dgraph-io/badger@v1.6.0/db.go:845 +0x618 fp=0xc000089ca0 sp=0xc000089ac8 pc=0xa2a828
github.com/dgraph-io/badger.(*DB).handleFlushTask(0xc0000fd180, 0xc0001aa060, 0x2a00000000, 0x39, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/runner/go/pkg/mod/github.com/dgraph-io/badger@v1.6.0/db.go:884 +0x5af fp=0xc000089e00 sp=0xc000089ca0 pc=0xa2ae8f
github.com/dgraph-io/badger.(*DB).flushMemtable(0xc0000fd180, 0xc00012a0a0, 0x0, 0x0)
	/home/runner/go/pkg/mod/github.com/dgraph-io/badger@v1.6.0/db.go:918 +0x1cc fp=0xc000089f98 sp=0xc000089e00 pc=0xa2b62c
github.com/dgraph-io/badger.Open.func4(0xc000010318)
	/home/runner/go/pkg/mod/github.com/dgraph-io/badger@v1.6.0/db.go:298 +0x8b fp=0xc000089fd8 sp=0xc000089f98 pc=0xa70e2b
runtime.goexit()
	/opt/hostedtoolcache/go/1.14.2/x64/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000089fe0 sp=0xc000089fd8 pc=0x47a871
created by github.com/dgraph-io/badger.Open
	/home/runner/go/pkg/mod/github.com/dgraph-io/badger@v1.6.0/db.go:297 +0x2308
```

https://github.com/ethersphere/bee/pull/75/checks?check_run_id=573849921#step:10:311

jmozah commented:
@janos fixed most of the comments by you. I am not sure about the dump though. Looks like it happened while opening the DB which i am doing so many times. May be this is related to CI machine... Lets see if it comes up again. I have raised a issue in badger repo also https://github.com/dgraph-io/badger/issues/1294

janos commented:
Thanks for addressing my comments, but now PutInBatch errors need to be handled https://github.com/ethersphere/bee/pull/75/checks?check_run_id=574344361#step:7:8.

jmozah commented:
@janos fixed the lints.. but the dump seems to be tricky for now. How do you want to proceed?

janos commented:
Thanks @jmozah. I think that we should not have panics related to pointer arithmetics, or actually, any type of panic. Would you mind to check if this problem is not present with badger 2.0.3?

jmozah commented:
2.0.3 also has the same issue... I dont think this is related to badger.. we are doing something wrong..

jmozah commented:
ok... looks like it works in 2.0.3...

janos commented:
> 2.0.3 also has the same issue... I dont think this is related to badger.. we are doing something wrong..

The panic is exactly pointing to badger using bbloom package code that handles pointers. It is clear what is the error.

jmozah commented:
> The panic is exactly pointing to badger using bbloom package code that handles pointers. It is clear what is the error.

They had some issues in the bloom filter i dont remember. @jarifibrahim  is looking in to it.

janos commented on go.mod:
```diff
@@ -5,7 +5,8 @@ go 1.14
 	github.com/coreos/go-semver v0.3.0
-	github.com/ethereum/go-ethereum v1.9.12
+	github.com/dgraph-io/badger/v2 v2.0.3
+	github.com/ethereum/go-ethereum v1.9.12 // indirect
```
We should not need go-ethereum anymore.

janos commented on pkg/shed/index.go:
```diff
@@ -162,32 +161,35 @@ func (f Index) Get(keyFields Item) (out Item, err error) {
+			if err == badger.ErrKeyNotFound {
+				return ErrNotFound
+			}
 			f.logger.Debugf("error getting key %s in Fill. Error: %s", string(key), err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		db.metrics.PutCount.Inc()
+		err = txn.Set(key, value)
+		if err != nil {
+			db.logger.Errorf("Could not insert chunk with key %s in DB", string(key))
```
Do not log error and return it.

janos replied:
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	})
+	if err != nil {
 		db.logger.Errorf("key %s not found in DB", string(key))
-		db.metrics.GetFailCounter.Inc()
+		db.logger.Debug("key %s not found in DB. Error : %v", string(key), err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-		db.logger.Debugf("encountered error during HAS of key %s. Error: %s ", string(key), err.Error())
-		db.metrics.HasFailCounter.Inc()
-		return false, err
+		db.logger.Errorf("key %s not found in DB", string(key))
+		db.logger.Debugf("key %s not found in DB. Error: %v", string(key), err.Error())
```
Do not log error and never return it fi the function should return it (unless there is an explicit reason for this). Just return it.

jarifibrahim commented:
We fixed something similar in badger v2.0.3. I'll look at v1.6.1 and get back. :+1:

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		db.metrics.DeleteCount.Inc()
+		err = txn.Delete(key)
+		if err != nil {
+			db.logger.Errorf("could not delete key %s from DB", string(key))
+			db.logger.Debugf("could not delete key %s from DB. Error: %v", string(key), err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while counting keys in DB")
+		db.logger.Debugf("error while doing count. Error : %w", err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while doing count. prefix : %v,", string(prefix))
+		db.logger.Debugf("error while doing count for prefix : %v, Error : %w", string(prefix), err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while doing iteration. startKey : %v, skipStartKey: %t",
+			string(startKey), skipStartKey)
+		db.logger.Debugf("error while doing iteration. startKey : %v, skipStartKey: %t, Error : %w",
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while finding first key with prefix : %v, Error : %w", string(prefix), err.Error())
+		db.logger.Debugf("error while doing first. prefix : %v, Error : %w", string(prefix), err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while finding last value for prefix : %v", string(prefix))
+		db.logger.Debugf("error while doing last. prefix : %v, Error : %w", string(prefix), err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +88,381 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
 	if err != nil {
-		db.logger.Debugf("could not writing batch. Error: %s ", err.Error())
-		db.metrics.WriteBatchFailCounter.Inc()
+		db.logger.Errorf("error while committing transaction. Error : %w", err.Error())
+		db.logger.Debugf("error while committing batch. Error : %w", err.Error())
```
Do not log error and return it.

janos replied:
Do not log error and return it.

jmozah commented:
> We fixed something similar in badger v2.0.3. I'll look at v1.6.1 and get back. 👍

@jarifibrahim  i took this as anopportunityy and moved to 2.0.3 :-)

jmozah commented:
@janos fixed all the comments. The build fails and it is not related to my changes.

janos commented:
@jmozah I ma sorry for no being clear enough. I think that no logging should be done related to errors if the error is returned. Could you please remove Errorf calls whenever you removed Debugf calls in the previous commit? They are not needed.

janos commented on pkg/shed/field_string.go:
```diff
@@ -65,6 +65,11 @@ func (f StringField) Put(val string) (err error) {
+func (f StringField) PutInBatch(batch *badger.Txn, val string) (err error) {
+	err = batch.Set(f.key, []byte(val))
+	if err != nil {
+		f.logger.Debugf("could not set values in PutInBatch. Error : %s", err.Error())
```
Do not log error and return it.

janos commented on pkg/shed/field_struct.go:
```diff
@@ -67,12 +67,16 @@ func (f StructField) Put(val interface{}) (err error) {
 	b, err := json.Marshal(val)
 	if err != nil {
-		f.logger.Debugf("could not PUT key %s in batch", string(f.key))
+		f.logger.Debugf("could not marshall while doing putiInBatch")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/field_struct.go:
```diff
@@ -67,12 +67,16 @@ func (f StructField) Put(val interface{}) (err error) {
+	}
+	err = batch.Set(f.key, b)
+	if err != nil {
+		f.logger.Debugf("could not put values in batch.")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/field_uint64.go:
```diff
@@ -67,20 +67,25 @@ func (f Uint64Field) Put(val uint64) (err error) {
+func (f Uint64Field) PutInBatch(batch *badger.Txn, val uint64) (err error){
+	err = batch.Set(f.key, encodeUint64(val))
+	if err != nil {
+		f.logger.Debugf("could not set uint64 value in PutInBatch",)
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/field_uint64.go:
```diff
@@ -67,20 +67,25 @@ func (f Uint64Field) Put(val uint64) (err error) {
 			val = 0
 		} else {
-			f.logger.Errorf("key %s not found. Error: %s", string(f.key), err.Error())
+			f.logger.Errorf("key %s not found", string(f.key))
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -141,17 +140,17 @@ func (db *DB) NewIndex(name string, funcs IndexFuncs, logger logging.Logger) (f
 	key, err := f.encodeKeyFunc(keyFields)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in Get. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in Get")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -141,17 +140,17 @@ func (db *DB) NewIndex(name string, funcs IndexFuncs, logger logging.Logger) (f
 	value, err := f.db.Get(key)
 	if err != nil {
-		f.logger.Debugf("error getting key %s in Get. Error: %s", string(key), err.Error())
+		f.logger.Debugf("error getting key %s in Get", string(key))
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -141,17 +140,17 @@ func (db *DB) NewIndex(name string, funcs IndexFuncs, logger logging.Logger) (f
 	out, err = f.decodeValueFunc(keyFields, value)
 	if err != nil {
-		f.logger.Debugf("error decofing keyfields in Get. Error: %s", err.Error())
+		f.logger.Debugf("error decofing keyfields in Get")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -198,7 +197,7 @@ func (f Index) Fill(items []Item) (err error) {
 	key, err := f.encodeKeyFunc(keyFields)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in Has. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in Has.")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -234,12 +240,12 @@ func (f Index) HasMulti(items ...Item) ([]bool, error) {
 	key, err := f.encodeKeyFunc(i)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in Put. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in Put")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -234,12 +240,12 @@ func (f Index) HasMulti(items ...Item) ([]bool, error) {
 	value, err := f.encodeValueFunc(i)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in Put. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in Put")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +254,50 @@ func (f Index) Put(i Item) (err error) {
 	key, err := f.encodeKeyFunc(i)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in PutInBatch. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in PutInBatch")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +254,50 @@ func (f Index) Put(i Item) (err error) {
 	value, err := f.encodeValueFunc(i)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in PutInBatch. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in PutInBatch")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +254,50 @@ func (f Index) Put(i Item) (err error) {
-	return nil
+	err = batch.Set(key, value)
+	if err != nil {
+		f.logger.Debugf("could not set values in batch")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +254,50 @@ func (f Index) Put(i Item) (err error) {
 	key, err := f.encodeKeyFunc(keyFields)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in Delete. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in Delete",)
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +254,50 @@ func (f Index) Put(i Item) (err error) {
 	key, err := f.encodeKeyFunc(keyFields)
 	if err != nil {
-		f.logger.Debugf("keyfields encoding error in DeleteInBatch. Error: %s", err.Error())
+		f.logger.Debugf("keyfields encoding error in DeleteInBatch")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +254,50 @@ func (f Index) Put(i Item) (err error) {
-	return nil
+	err = batch.Delete(key)
+	if err != nil {
+		f.logger.Debugf("could not delete items in batch")
```
Error is returned, this log line is not needed.

janos commented on pkg/shed/vector_uint64.go:
```diff
@@ -66,16 +66,21 @@ func (f Uint64Vector) Put(i, val uint64) (err error) {
+func (f Uint64Vector) PutInBatch(batch *badger.Txn, i, val uint64) (err error){
+	err = batch.Set(f.indexKey(i), encodeUint64(val))
+	if err != nil {
+		f.logger.Debugf("could not put uint64 value in batch")
```
Error is returned, this log line is not needed. Just return the error from batch.Set.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		}
+		return item.Value(func(val []byte) error {
+			value = val
+			db.logger.Tracef("got valeu for key %s with length %d", string(key), len(val))
```
valeu -> value

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		})
+	})
+	if err != nil {
 		db.logger.Errorf("key %s not found in DB", string(key))
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-		db.logger.Debugf("encountered error during HAS of key %s. Error: %s ", string(key), err.Error())
-		db.metrics.HasFailCounter.Inc()
-		return false, err
+		db.logger.Errorf("key %s not found in DB", string(key))
```
This log line is not needed.

The error is not returned at all please return it, and remove the log line.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		db.metrics.DeleteCount.Inc()
+		err = txn.Delete(key)
+		if err != nil {
+			db.logger.Errorf("could not delete key %s from DB", string(key))
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while counting keys in DB")
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while doing count. prefix : %v,", string(prefix))
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error doing count from. prefix : %v", string(prefix))
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while doing iteration. startKey : %v, skipStartKey: %t",
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while finding first key with prefix : %v", string(prefix))
```
This log line is not needed, as the error is returned.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while finding last value for prefix : %v", string(prefix))
```
This log line is not needed, as the error is returned.

janos replied:
This comment left unaddressed.

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,372 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
 	if err != nil {
-		db.logger.Debugf("could not writing batch. Error: %s ", err.Error())
-		db.metrics.WriteBatchFailCounter.Inc()
+		db.logger.Errorf("error while committing transaction",)
```
This log line is not needed, as the error is returned.

jmozah commented:
@janos coding guidelines says "This means that the same problematic event may have two log lines but with different severities" thats why i had Errorf and Debugf messages. This is still not clear for me where to put Error and where to put debug and where to put both.

janos commented:
@jmozah It also states in https://github.com/ethersphere/bee/blob/master/CODING.md#error-propagation-annotation-and-handling section that errors should not be logged if they are returned. `Error should not be logged and passed at the same time.` That is the point of my comments. Sorry if that was not clear.

jmozah commented:
@janos no issues. it is clear now for me. I made the changes.

janos commented:
> @janos no issues. it is clear now for me. I made the changes.

Thanks @jmozah

janos commented on pkg/shed/field_string.go:
```diff
@@ -65,6 +65,10 @@ func (f StringField) Put(val string) (err error) {
-func (f StringField) PutInBatch(batch *leveldb.Batch, val string) {
-	batch.Put(f.key, []byte(val))
+func (f StringField) PutInBatch(batch *badger.Txn, val string) (err error) {
+	err = batch.Set(f.key, []byte(val))
+	if err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	return batch.Set(f.key, []byte(val))
```

janos commented on pkg/shed/field_uint64.go:
```diff
@@ -67,20 +67,23 @@ func (f Uint64Field) Put(val uint64) (err error) {
-func (f Uint64Field) PutInBatch(batch *leveldb.Batch, val uint64) {
-	batch.Put(f.key, encodeUint64(val))
+func (f Uint64Field) PutInBatch(batch *badger.Txn, val uint64) (err error){
+	err = batch.Set(f.key, encodeUint64(val))
+	if err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	return batch.Set(f.key, encodeUint64(val))
```

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +248,44 @@ func (f Index) Put(i Item) (err error) {
 	}
-	batch.Put(key, value)
-	return nil
+	err = batch.Set(key, value)
+	if err != nil {
+		return err
+	}
+	return err
```
```suggestion
	return batch.Set(key, value)
```

janos commented on pkg/shed/index.go:
```diff
@@ -248,42 +248,44 @@ func (f Index) Put(i Item) (err error) {
 	}
-	batch.Delete(key)
-	return nil
+	err = batch.Delete(key)
+	if err != nil {
+		return err
+	}
+	return err
```
```suggestion
	return batch.Delete(key)
```

janos commented on pkg/shed/vector_uint64.go:
```diff
@@ -66,16 +66,20 @@ func (f Uint64Vector) Put(i, val uint64) (err error) {
-func (f Uint64Vector) PutInBatch(batch *leveldb.Batch, i, val uint64) {
-	batch.Put(f.indexKey(i), encodeUint64(val))
+func (f Uint64Vector) PutInBatch(batch *badger.Txn, i, val uint64) (err error){
+	err = batch.Set(f.indexKey(i), encodeUint64(val))
+	if err != nil {
+		return err
+	}
+	return err
```
```suggestion
	return batch.Set(f.indexKey(i), encodeUint64(val))
```

janos commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
 	if err != nil {
-		db.logger.Debugf("encountered error during HAS of key %s. Error: %s ", string(key), err.Error())
-		db.metrics.HasFailCounter.Inc()
-		return false, err
+		db.metrics.HasFailCount.Inc()
+	}
```
The error from line 144 is never returned. I think that it should be in this if block.

acud commented on go.sum:
```diff
@@ -95,10 +99,17 @@ github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea h1:j4317fAZh7X
 github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea/go.mod h1:93vsz/8Wt4joVM7c2AVqh+YRMiUSc14yDtF28KmMOgQ=
 github.com/dgraph-io/badger v1.5.5-0.20190226225317-8115aed38f8f/go.mod h1:VZxzAIRPHRVNRKRo6AXrX9BJegn6il06VMTZVJYCIjQ=
 github.com/dgraph-io/badger v1.6.0-rc1/go.mod h1:zwt7syl517jmP8s94KqSxTlM6IMsdhYy6psNgSztDR4=
+github.com/dgraph-io/badger v1.6.0 h1:DshxFxZWXUcO0xX476VJC07Xsr6ZCBVRHKZ93Oh7Evo=
```
i think this version is not needed anymore...

janos replied:
It looks like that we still depend on swarm/storage in shed example, so tis sum is added here. We should remove completely swarm dependencies in the future PRs. And then this line could be removed then.

acud commented on go.sum:
```diff
@@ -109,14 +120,17 @@ github.com/docker/go-connections v0.4.0/go.mod h1:Gbd7IOopHjR8Iph03tsViu4nIes5Xh
 github.com/elazarl/goproxy v0.0.0-20170405201442-c4fc26588b6e/go.mod h1:/Zj4wYkgs4iZTTu3o/KG3Itv/qCCa8VVMlb3i9OVuzc=
 github.com/envoyproxy/go-control-plane v0.9.1-0.20191026205805-5f8ba28d4473/go.mod h1:YTl/9mNaCwkRvm6d1a2C3ymFceY/DCBVvsKhRF0iEA4=
 github.com/envoyproxy/protoc-gen-validate v0.1.0/go.mod h1:iSmxcyjqTsJpI2R4NaDN7+kN2VEUnK/pcBlmesArF7c=
+github.com/ethereum/go-ethereum v1.9.2 h1:RMIHDO/diqXEgORSVzYx8xW9x2+S32PoAX5lQwya0Lw=
```
this go ethereum stuff probably not too

janos replied:
It looks like that we still depend on swarm/storage in shed example, so tis sum is added here. We should remove completely swarm dependencies in the future PRs. And then this line could be removed then. Now go tool keeps adding it.

acud commented on pkg/shed/db.go:
```diff
@@ -23,45 +23,58 @@
+	o.Logger = nil // Dont enable the badger logs
+	database, err := badger.Open(o)
 	if err != nil {
+		logger.Error("could not open database.")
```
not needed to log _and_ return the error. enough just to return the error in this case

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	db.quit = make(chan struct{})
-
+	db.metrics.DBOpenCount.Inc()
+	logger.Debugf("Opened DB with path : %s, valueThreshold : %d, valueLogMaxEntries : %d",
```
I don't think this is needed

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	// Create a quit channel for the periodic metrics collector and run it
-	db.quit = make(chan struct{})
-
+	db.metrics.DBOpenCount.Inc()
```
not sure if this metric is all that useful. this should always be 1, at least for the time being...

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	}
-	db.metrics.PutCounter.Inc()
-	return nil
+	db.logger.Tracef("Put value len : %d for key %s", len(value), string(key))
```
this will flood the logs on `trace` (imagine thousands of these lines when you upload and sync). i think this is not needed

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+			db.metrics.PutFailCount.Inc()
+			return err
+		}
+		db.logger.Tracef(" put success with key %s and value len %d", string(key), len(value))
```
here too

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	} else {
-		db.logger.Errorf("key %s not found in DB", string(key))
-		db.metrics.GetFailCounter.Inc()
+	db.logger.Tracef("Get value for key %s", string(key))
```
again

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		})
+	})
+	if err == nil {
+		db.logger.Tracef("got value with len : %d for key : %s", len(value), string(key))
```
i dont think this is needed. this will flood

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+// it returns a bool indicating true or false OR error if it encounters one during the operation.
 func (db *DB) Has(key []byte) (yes bool, err error) {
-	yes, err = db.ldb.Has(key, nil)
+	db.logger.Tracef("Has value with key %s", string(key))
```
ditto

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	db.metrics.HasCounter.Inc()
+
+	if yes {
+		db.logger.Tracef("found key %s", string(key))
```
ditto

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	if yes {
+		db.logger.Tracef("found key %s", string(key))
+	} else {
+		db.logger.Tracef("could not found key %s", string(key))
```
ditto

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-	if err != nil {
-		db.logger.Debugf("could not DELETE key %s. Error: %s ", string(key), err.Error())
-		db.metrics.DeleteFailCounter.Inc()
+	db.logger.Tracef("deleting key %s along with value", string(key))
```
also

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+		if err != nil {
+			db.metrics.DeleteFailCount.Inc()
+		} else {
+			db.logger.Tracef("deleted key %s", string(key))
```
you get the point

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+
+// Count gives a count of all the keys present in the DB.
+func (db *DB) Count(ctx context.Context) (count int, err error) {
+	db.logger.Tracef("counting all keys")
```
not needed

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	if err != nil {
+		db.metrics.TotalFailCount.Inc()
+	} else {
+		db.logger.Tracef("Total count is %d", count)
```
i dont think this is needed when the count is returned to the calling context

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+// a nil prefix acts like the total count of the DB
+func (db *DB) CountPrefix(prefix []byte) (count int, err error) {
+	if prefix != nil {
+		db.logger.Tracef("counting all keys with prefix %s", string(prefix))
```
not sure if this is so helpful

acud commented on pkg/shed/db.go:
```diff
@@ -74,83 +87,359 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	if err != nil {
+		db.metrics.CountPrefixFailCount.Inc()
+	} else {
+		db.logger.Tracef("Total count for prefix %s is %d", string(prefix), count)
```
again

jmozah commented:
Bee is still dependent on Swarm and Ethereum as the example_store_test.go in shed uses it.
Once we move chun related stuff inside bee.. we will get rid of Swarm and ethereum.

acud commented on go.mod:
```diff
@@ -1,33 +1,33 @@
 module github.com/ethersphere/bee
 
-go 1.14
+go 1.13
```
should be `1.14`

acud commented on pkg/shed/index.go:
```diff
@@ -162,32 +158,32 @@ func (f Index) Get(keyFields Item) (out Item, err error) {
+			if err == badger.ErrKeyNotFound {
+				return ErrNotFound
+			}
+			f.logger.Debugf("error getting key %s in Fill.", string(key))
```
can also be removed, since error is returned

acud commented on pkg/shed/vector_uint64.go:
```diff
@@ -127,10 +130,10 @@ func (f Uint64Vector) Dec(i uint64) (val uint64, err error) {
+		if err == ErrNotFound {
 			val = 0
 		} else {
 			f.logger.Debugf("error getiing value while doing DecInBatch. Error: %s", err.Error())
```
this too is not necessary since error is returned

acud commented on pkg/shed/vector_uint64.go:
```diff
@@ -89,18 +89,21 @@ func (f Uint64Vector) Inc(i uint64) (val uint64, err error) {
+		if err == ErrNotFound {
 			val = 0
 		} else {
 			f.logger.Debugf("error getiing value while doing IncInBatch. Error: %s", err.Error())
```
same

acud commented on pkg/shed/vector_uint64.go:
```diff
@@ -66,16 +66,16 @@ func (f Uint64Vector) Put(i, val uint64) (err error) {
+		if err == ErrNotFound {
 			val = 0
 		} else {
 			f.logger.Debugf("error getiing value while doing Inc. Error: %s", err.Error())
```
also

acud commented on pkg/shed/index.go:
```diff
@@ -322,102 +316,66 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
+	// create a copy of value byte slice not to share badger underlaying slice array
+	valueItem, err := f.decodeValueFunc(keyItem, append([]byte(nil), value...))
 	if err != nil {
 		f.logger.Debugf("error decoding value in itemFromIterator. Error: %s", err.Error())
```
error returned, no need to log

acud commented on pkg/shed/index.go:
```diff
@@ -322,102 +316,66 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
+	// create a copy of key byte slice not to share badger underlaying slice array
 	keyItem, err := f.decodeKeyFunc(append([]byte(nil), key...))
 	if err != nil {
 		f.logger.Debugf("error decoding key in itemFromIterator. Error: %s", err.Error())
```
error returned, no need to log

acud commented on pkg/shed/field_uint64.go:
```diff
@@ -131,10 +133,10 @@ func (f Uint64Field) Dec() (val uint64, err error) {
 	if err != nil {
-		if err == leveldb.ErrNotFound {
+		if err == ErrNotFound {
 			f.logger.Debugf("key %s not found", string(f.key))
```
not necessary probably

acud commented on pkg/shed/field_uint64.go:
```diff
@@ -131,10 +133,10 @@ func (f Uint64Field) Dec() (val uint64, err error) {
+		if err == ErrNotFound {
 			f.logger.Debugf("key %s not found", string(f.key))
 			val = 0
 		} else {
```
one line below, error returned no need to log

acud commented on pkg/shed/field_uint64.go:
```diff
@@ -113,7 +115,7 @@ func (f Uint64Field) IncInBatch(batch *leveldb.Batch) (val uint64, err error) {
 	if err != nil {
-		if err == leveldb.ErrNotFound {
+		if err == ErrNotFound {
 			f.logger.Debugf("key %s not found", string(f.key))
```
same, same below

acud commented on pkg/shed/field_uint64.go:
```diff
@@ -91,10 +90,10 @@ func (f Uint64Field) Inc() (val uint64, err error) {
 	if err != nil {
-		if err == leveldb.ErrNotFound {
+		if err == ErrNotFound {
 			f.logger.Debugf("key %s not found", string(f.key))
```
no need to log here i think

acud commented on pkg/shed/field_string.go:
```diff
@@ -49,7 +49,7 @@ func (db *DB) NewStringField(name string, logger logging.Logger) (f StringField,
 	if err != nil {
-		if err == leveldb.ErrNotFound {
+		if err == ErrNotFound {
 			f.logger.Errorf("key %s not found", string(f.key))
```
if there's an error, you would would probably want to return it no? here you log it, but then return an empty string but nil error. i'm not sure this is right

jmozah replied:
ok. The idea was to split the error in to KeyNotFound and rest. because the application might want to respond differently... but since you guys seems to like the idea of just returning the err as it is, i am fixing in all those places.

acud commented on pkg/shed/db.go:
```diff
@@ -73,84 +85,307 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
-// Close closes LevelDB database.
+// Close shuts down the badger DB.
 func (db *DB) Close() (err error) {
-	close(db.quit)
```
i wonder, we dont need this anymore?

jmozah replied:
Where is this?   we have a Close() at the end of the file which doesn't look like this.

```
// Close shuts down the badger DB.
func (db *DB) Close() (err error) {
	db.logger.Tracef("database closed with path %s", db.path)
	return db.bdb.Close()
}
```

acud commented on pkg/shed/field_string_test.go:
```diff
@@ -37,7 +37,7 @@ func TestStringField(t *testing.T) {
 		got, err := simpleString.Get()
-		if err != nil {
+		if err == nil {
 			t.Fatal(err)
```
you'd probably want to write a custom error message here for the test, since you're failing the test with a `nil` value.

acud replied:
I'm merging the PR as is, you can piggy-back this change on a later PR

jmozah replied:
Thanks @acud
